source/common/http/http2/codec_impl.cc

```diff
@@ -543,6 +577,57 @@ int ConnectionImpl::onStreamClose(int32_t stream_id, uint32_t error_code) {
   return 0;
 }
 
+int ConnectionImpl::onMetadataReceived(int32_t stream_id, const uint8_t* data, size_t len) {
+  ENVOY_CONN_LOG(trace, "recv {} bytes METADATA", connection_, len);
+
+  StreamImpl* stream = getStream(stream_id);
+  if (!stream) {
+    return 0;
+  }
+
+  bool success = stream->getMetadataDecoder().receiveMetadata(data, len);
+  return success ? 0 : NGHTTP2_ERR_CALLBACK_FAILURE;
+}
+
+int ConnectionImpl::onMetadataFrameComplete(int32_t stream_id, bool end_metadata) {
+  ENVOY_CONN_LOG(trace, "recv METADATA frame on stream {}, end_metadata: {}", connection_,
+                 stream_id, end_metadata);
+
+  StreamImpl* stream = getStream(stream_id);
+  if (!stream) {
+    return 0;
+  }
+  bool result = stream->getMetadataDecoder().onMetadataFrameComplete(end_metadata);
+  return result ? 0 : NGHTTP2_ERR_CALLBACK_FAILURE;
+}
+
+ssize_t ConnectionImpl::packMetadata(int32_t stream_id, uint8_t* buf, size_t len) {
+  ENVOY_CONN_LOG(trace, "pack METADATA frame on stream {}", connection_, stream_id);
+
+  StreamImpl* stream = getStream(stream_id);
+  if (!stream) {
+    return 0;
+  }
+
+  MetadataEncoder& encoder = stream->getMetadataEncoder();
+  const uint64_t size_to_copy = std::min(METADATA_MAX_PAYLOAD_SIZE, encoder.payload().length());
+  // nghttp2 guarantees len is at least 16KiB.
+  ASSERT(len >= size_to_copy);
+
+  Buffer::OwnedImpl& p = reinterpret_cast<Buffer::OwnedImpl&>(encoder.payload());
+  p.copyOut(0, size_to_copy, buf);
+
+  // Releases the payload that has been copied to nghttp2.
+  encoder.releasePayload(size_to_copy);
+
+  // Keep submitting extension frames if there is payload left in the encoder.
+  if (encoder.hasNextFrame()) {
```

Comment:
so when the caller calls encodeMetadata with a large frame, encodeMetadata calls submitMetadata, submitMetadata results in nghttp2 calling packMetadata, which may result in another submitMetadata?

Any chance we can remove the recursion by moving where we check for hasNextFrame and calling submitMetadata multiple times instead?

Reply:
You are right. The old code let packMetadata call submitMetadata until there is no metadata left. I can see it is very hard to understand. Changed according to the suggestion.